### PR TITLE
SAPI5 engine fixes

### PR DESCRIPTION
**Pull request by daanzu**

### CODE REVIEW

File: dragonfly/engines/backend_sapi5/engine.py [hunk 1]
```diff
@@ -221,8 +221,12 @@ def set_exclusiveness(self, grammar, exclusive):
...
                         % (grammar.name, exclusive))
         grammar_handle = self._get_grammar_wrapper(grammar).handle
-        grammar_handle.State = constants.SGSExclusive
-       # grammar_handle.SetGrammarState(constants.SPGS_EXCLUSIVE)
+        if exclusive:
+            grammar_handle.State = constants.SGSExclusive
+        else:
+            grammar_handle.State = constants.SGSEnabled
+            # Should we set this to SGSDisabled if grammar is not enabled or was deactivated before being set exclusive?
+        # grammar_handle.SetGrammarState(constants.SPGS_EXCLUSIVE)
```
Comment (Danesprite) [inline dragonfly/engines/backend_sapi5/engine.py]: Yes, I think that makes sense.

**Comment by Danesprite** [comment]
Thanks!

Did you want to add the sapi5inproc sleep/wake grammar you mentioned? It sounded pretty useful.

**Comment by daanzu** [comment]
Updated. Perhaps we should add a pointer to the loader somewhere in documentation for getting started? Not sure where.

**Comment by Danesprite** [comment]
Great!

I had no idea those wave files existed under *C:\Windows\Media*.

The new loader should be mentioned in the [Installation section for WSR](https://dragonfly2.readthedocs.io/en/latest/installation.html#installation-for-windows-speech-recognition). You can make a new section on your Kaldi engine in there if you like.

**Comment by daanzu** [comment]
Updated. Should be done.

**Comment by Danesprite** [comment]
Looks good.

Don't know if you've noticed this happening sometimes:
```
engine (ERROR): Grammar _recobs_grammar: exception: Recognition observer received an unexpected recognition: (u'no', u'can', u'do')
Traceback (most recent call last):
  File "w:\dragonfly-daanzu\dragonfly\engines\backend_sapi5\engine.py", line 630, in recognition_callback
    if not func(words):
  File "w:\dragonfly-daanzu\dragonfly\engines\backend_sapi5\recobs.py", line 72, in process_recognition
    " recognition: %s" % (words,))
RuntimeError: Recognition observer received an unexpected recognition: (u'no', u'can', u'do')
engine (ERROR): Grammar _recobs_grammar: failed to decode recognition [u'no', u'can', u'do'].
```

It seems like "no can do" is being recognised. Perhaps it could be fixed by using a less likely transition for `Impossible` elements in the compiler? E.g. `"impossible " * 10` instead of `"no can do"`? Or maybe there's a better way?

https://github.com/dictation-toolbox/dragonfly/blob/68aa12183a0000d61b4d963b6835db34e4c20861/dragonfly/engines/backend_sapi5/compiler.py#L221-L223

**Comment by daanzu** [comment]
I don't really use recognition observers, so I haven't seen that happen. When does it happen? I had noticed that in the code a while back and also thought it seemed far less than "impossible". I wouldn't think there's much of a downside to your proposed change. Want me to go ahead and put it in?

**Comment by Danesprite** [comment]
It happened a few times when I was testing the module loaders.

Thinking about it, a better solution is to compile `Impossible` elements as `RuleRefs` using the fake rule already added to each grammar in `compile_grammar()`. I'm guessing this is what t4ngo was thinking of doing originally.

I'll add & test that. Should be ready to merge afterwards.

**Comment by Danesprite** [comment]
Done.

Probably not a big deal, but I just tried running the test suite and the exclusiveness test failed:

```
FAIL: test_exclusive_grammars (dragonfly.test.test_contexts.TestRules)
Verify that the engine supports exclusive grammars.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "w:\dragonfly-daanzu\dragonfly\test\test_contexts.py", line 99, in test_exclusive_grammars
    grammar1.set_exclusiveness(False)
  File "w:\dragonfly-daanzu\dragonfly\grammar\grammar_base.py", line 156, in set_exclusiveness
    self._engine.set_exclusiveness(self, exclusive)
  File "w:\dragonfly-daanzu\dragonfly\engines\backend_sapi5\engine.py", line 228, in set_exclusiveness
    assert wrapper.state_before_exclusive in (constants.SGSEnabled, constants.SGSDisabled)
AssertionError
```

**Comment by daanzu** [comment]
Hmm, I'll take a look at exclusiveness.

**Comment by Danesprite** [comment]
I've figured it out. The problem is not with the engine, but with the tests themselves. There are a few problems with them that affect SAPI5 and other engines. I think this should be merged now and I can fix those issues later.

**Comment by daanzu** [comment]
Ah, cool; I hadn't gotten around to investigating it yet. Merging sounds good to me.

**Comment by Danesprite** [comment]
Okay then. I'll merge this now.

Exclusiveness and contexts are difficult to test with natlink because DNS will normally just try to type mimicked words into the foreground window if they don't match an active rule. I didn't catch this until now because of how `NatlinkEngine.mimic()` behaves. I'll exclude these tests from the natlink suite at least. `Grammar.set_exclusiveness()` is used indirectly by most of dragonfly's tests in any case.

**Pull request merged**